libraries/SPI/SPI.cpp

```diff
@@ -311,53 +319,55 @@ void SPIClass::write32(uint32_t data, bool msb) {
 
 /**
  * Note:
- *  data need to be aligned to 32Bit
- *  or you get an Fatal exception (9)
+ *  data needs to be aligned to 32 bit boundaries like this: char outBuffer[spiBufferSize] __attribute__ ((aligned (4)));
+ *  if not you may get an Fatal exception (9)
  * @param data uint8_t *
  * @param size uint32_t
  */
 void SPIClass::writeBytes(uint8_t * data, uint32_t size) {
+	uint32_t localSize;
     while(size) {
-        if(size > 64) {
-            writeBytes_(data, 64);
-            size -= 64;
-            data += 64;
-        } else {
-            writeBytes_(data, size);
-            size = 0;
-        }
-    }
+        if(size > 64)
+			localSize = 64;
+		else
+			localSize = size;
+			
+        writeBytes_(data, localSize);
+        size -= localSize;
+		if ((size != 0) && (data)) data += localSize;
+	}
 }
 
-void SPIClass::writeBytes_(uint8_t * data, uint8_t size) {
+
+inline void SPIClass::writeBytes_(uint8_t * data, uint32_t size) {
+#ifndef SPI_SKIP_INITIAL_BUSY_CHECK
     while(SPI1CMD & SPIBUSY) {}
-    // Set Bits to transfer
+#endif
+    // Set bits to transfer
     setDataBits(size * 8);
 
-    volatile uint32_t * fifoPtr = &SPI1W0;
-    uint32_t * dataPtr = (uint32_t*) data;
-    uint8_t dataSize = ((size + 3) / 4);
-
-    while(dataSize--) {
-        *fifoPtr = *dataPtr;
-        dataPtr++;
-        fifoPtr++;
-    }
-
+	if(data) {
+		memcpy((uint8_t*)&SPI1W0, data, size);
```

Comment:
In general, accessing *most* peripheral registers other than with `l32i`/`s32i` isn't valid, but if you use memcpy with an unaligned source buffer this may very well happen.
`memcpy` is indeed better than a plain loop, because on Xtensa it interleaves multiple loads and stores, reducing pipeline stalls due to memory/APB access. I suggest to have a check whether `data` is 32-bit aligned first. If it is, then go via "fast path" using `memcpy`. If it's not, go via "slow path" of a loop with one memcpy per word. Unlike the old code this solution will gracefully handle unaligned source buffers, removing the alignment restriction (and for aligned buffers we will gain performance due to `memcpy`).